### ISSUE TITLE
Out Of Memory error when processing a large file

### ISSUE THREAD

**Issue by Hydragyrum** [Mermade/widdershins#273]
**Describe the bug**
A clear and concise description of what the bug is. Reports containing a full reproduction recipe (command-line and input file gist) will get a **much** faster response.

Hello,

I have an OpenApi v3.0 spec generated by Swagger 2.1.0 via a lot of java code. The resulting YAML file is around 19 MB long.  For confidentiality reasons, I am unable to provide this file.

Spectacle is able to generate something out of it, but widdershins crashes which an out of memory error. The crash log follows

```
Compiling all doT templates...
Loaded def authentication.def
Loaded def authentication_none.def
Loaded def callbacks.def
Loaded def debug.def
Loaded def discovery.def
Loaded def footer.def
Loaded def links.def
Loaded def parameters.def
Loaded def responses.def
Loaded def security.def
Compiling code_go.dot to function
Compiling code_http.dot to function
Compiling code_java.dot to function
Compiling code_javascript.dot to function
Compiling code_nodejs.dot to function
Compiling code_php.dot to function
Compiling code_python.dot to function
Compiling code_ruby.dot to function
Compiling code_shell.dot to function
Compiling main.dot to function
Compiling operation.dot to function
Compiling translations.dot to function

<--- Last few GCs --->

[9635:0x5321c80]    90616 ms: Scavenge 2036.3 (2053.9) -> 2031.6 (2067.9) MB, 12.2 / 0.0 ms  (average mu = 0.163, current mu = 0.149) allocation failure


<--- JS stacktrace --->

==== JS stack trace =========================================

    0: ExitFrame [pc: 0x1391439]
Security context: 0x311a5f3808d1 <JSObject>
    1: traverse(aka traverse) [0x1d7c65f4afd1] [/home/hydra/.nvm/versions/node/v13.1.0/lib/node_modules/widdershins/node_modules/openapi-sampler/dist/openapi-sampler.js:~740] [pc=0x9d89818abe6](this=0x003de95804b1 <undefined>,0x08c42ac040b9 <Object map = 0x3637711aa969>,0x1d3f563bf799 <Object map = 0x1ba927f61729>,0x0599760e87e9 <Object map = 0x36e5671658c9>)
   ...

FATAL ERROR: Ineffective mark-compacts near heap limit Allocation failed - JavaScript heap out of memory
 1: 0x9e9f40 node::Abort() [node]
 2: 0x9ec192 node::OnFatalError(char const*, char const*) [node]
 3: 0xb4611e v8::Utils::ReportOOMFailure(v8::internal::Isolate*, char const*, bool) [node]
 4: 0xb46499 v8::internal::V8::FatalProcessOutOfMemory(v8::internal::Isolate*, char const*, bool) [node]
 5: 0xcf3535  [node]
 6: 0xcf3bc6 v8::internal::Heap::RecomputeLimits(v8::internal::GarbageCollector) [node]
 7: 0xd003fa v8::internal::Heap::PerformGarbageCollection(v8::internal::GarbageCollector, v8::GCCallbackFlags) [node]
 8: 0xd01305 v8::internal::Heap::CollectGarbage(v8::internal::AllocationSpace, v8::internal::GarbageCollectionReason, v8::GCCallbackFlags) [node]
 9: 0xd0299f v8::internal::Heap::HandleGCRequest() [node]
10: 0xcb1505 v8::internal::StackGuard::HandleInterrupts() [node]
11: 0x100e863 v8::internal::Runtime_StackGuard(int, unsigned long*, v8::internal::Isolate*) [node]
12: 0x1391439  [node]
[1]    9635 abort (core dumped)  widdershins -v <filename>.yaml -o
```

The parsing of the file seems to have worked, at it was able to dereference the schemas properly, However, as soon as the templates were compiled, memory usage spiked to over 2 GB before core dumping.

* [x] - I have checked that my input document is valid OpenAPI 2.0/3.0.x or AsyncAPI 1.x

**To Reproduce**
1. Command-line:  `widdershins -v <filename>.yaml -o <outputFile>.md`
2. My input file: (please link to a gist)
For confidentiality reasons, I am unable to link the input file :(
The file has approximately 3100 lines describing the services, and approximately 310000 lines describing components and schemas, which may be part of what's causing the problems.

**Expected behavior**

I expect the markdown file to be properly generated.

**Screenshots**

**Versions**:
- Node.js (note only LTS versions are supported): 12.13.1

**Additional context**

**Comment by Hydragyrum** [comment]
Looking more closely, it's possible that circular references are causing the problem, as it seems to be trying to systematically expand everything

**Comment by jasoncoding** [comment]
I am having this exact problem with a json input file of only 170k.  My output:

```
Compiling all doT templates...
Loaded def authentication.def
Loaded def authentication_none.def
Loaded def discovery.def
Loaded def footer.def
Loaded def links.def
Loaded def parameters.def
Loaded def responses.def
Loaded def security.def
Compiling code_go.dot to function
Compiling code_http.dot to function
Compiling code_java.dot to function
Compiling code_javascript.dot to function
Compiling code_nodejs.dot to function
Compiling code_php.dot to function
Compiling code_python.dot to function
Compiling code_ruby.dot to function
Compiling code_shell.dot to function
Compiling main.dot to function
Compiling operation.dot to function
Compiling translations.dot to function
# sampler Maximum call stack size exceeded

<--- Last few GCs --->

[21636:000002F9CDF5E730]    13734 ms: Mark-sweep 1397.8 (1425.8) -> 1397.2 (1424.8) MB, 768.5 / 0.0 ms  (average mu = 0.043, current mu = 0.012) allocation failure scavenge might not succeed
[21636:000002F9CDF5E730]    13737 ms: Scavenge 1398.0 (1424.8) -> 1397.4 (1425.3) MB, 1.8 / 0.0 ms  (average mu = 0.043, current mu = 0.012) allocation failure


<--- JS stacktrace --->

==== JS stack trace =========================================

    0: ExitFrame [pc: 0000030A844DC5C1]
Security context: 0x01ef62d1e6e1 <JSObject>
    1: keys [000001EF62D05199](this=0x01ef62d04631 <JSFunction Object (sfi = 0000037AABA0E331)>,0x01d356fe1e81 <Object map = 000000029E0B5529>)
    2: visit [000000F0606C99F9] [C:\Users\jason\AppData\Roaming\npm\node_modules\widdershins\node_modules\safe-json-stringify\index.js:~23] [pc=0000030A84CD733C](this=0x026d1a61ad49 <JSGlobal Object>,obj=0x01d356fe...

FATAL ERROR: Ineffective mark-compacts near heap limit Allocation failed - JavaScript heap out of memory
```

**Comment by Hydragyrum** [comment]
@jasoncoding do you have something in your schema that links back to itself?

**Comment by jasoncoding** [comment]
@Hydragyrum thanks for the tip, I shall try that.

**Comment by MikeRalphson** [comment]
Should be fixed in version `4.0.0`, which includes breaking changes (please see the `README.md` file).
